pkg/analytics/categories.go

```diff
@@ -0,0 +1,127 @@
+package analytics
+
+import (
+	"slices"
+	"strings"
+)
+
+type CliArgCategorizer struct {
+	commands []string // parsed commands
+	flags    []string // parsed flags
+	result   []string // final categories passed to output
+
+	flagsWhitelist  []string
+	knownCommands   []string
+	knownProducts   []string // {"sbom", "code", "iac", "container"}
+	productFallback string   // {"oss"}
+}
+
+// NewCategoryParser creates a new instance of CliArgCategorizer, which is used to
+// parse and categorize command-line arguments based on known commands, whitelisted flags,
+// and recognized products.
+//
+// Parameters:
+//   - knownCommands: A slice of strings representing valid main commands and subcommands.
+//   - flagsWhitelist: A slice of strings representing whitelisted flags (without leading dashes).
+//   - knownProducts: A slice of strings representing the names of recognized products.
+//
+// Returns:
+//   - A pointer to a new CliArgCategorizer instance, ready to be used for parsing arguments.
+//
+// Usage:
+// categorizer := NewCategoryParser(knownCommands, whitelistedFlags, knownProducts)
+// categoryVector = categorizer.CategorizeCliArgs(os.Args)
+func NewCategoryParser(knownCommands []string, flagsWhitelist []string, knownProducts []string) *CliArgCategorizer {
+	return &CliArgCategorizer{
+		commands: []string{},
+		flags:    []string{},
+		result:   []string{},
```

Comment:
Suggestion: caching results like this introduces a state to the structure, this makes it impossible to use a parser concurrently. This seems an unnecessary limitation. Just use parameters instead.